### PR TITLE
gst123: init at 0.4.1

### PR DESCRIPTION
**Pull request by swesterfeld**
## Description of changes

GStreamer based command line media player

The program gst123 is designed to be a more flexible command line player in
the spirit of ogg123 and mpg123, based on GStreamer. It plays all file formats
supported by GStreamer, so if you have audio/video collections which contain
different file formats, like flac, ogg and mp3, you can use gst123 to play all
your audio/video files.

https://space.twc.de/~stefan/gst123.php

Other distributions (like Ubuntu) package gst123 since many years. I am the upstream developer.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -16266,6 +16266,12 @@
     github = "sweenu";
     githubId = 7051978;
   };
+  swesterfeld = {
```
Comment (AndersonTorres) [inline maintainers/maintainer-list.nix:16269]: Split this in a separate commit - no need to open a PR.

File: pkgs/applications/audio/gst123/default.nix [hunk 2]
```diff
@@ -0,0 +1,28 @@
+{ lib, stdenv, fetchurl, pkg-config, wrapGAppsHook, gst_all_1, gtk3, ncurses }:
```
Comment (AndersonTorres) [inline pkgs/applications/audio/gst123/default.nix:1]: One per line:
```suggestion
{ lib
, stdenv
, fetchurl
, pkg-config
, wrapGAppsHook
, gst_all_1
, gtk3
, ncurses
}:
```

File: pkgs/applications/audio/gst123/default.nix [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+stdenv.mkDerivation rec {
+  pname = "gst123";
+  version = "0.4.0";
+  src = fetchurl {
+    url = "https://github.com/swesterfeld/gst123/releases/download/${version}/gst123-${version}.tar.bz2";
+    sha256 = "sha256-ThE7Us1ZGNjrsG8DkdL3NXk9s6bcW6F3+krghRnJzAQ=";
+  };
```
Comment (AndersonTorres) [inline pkgs/applications/audio/gst123/default.nix:9]: Use this:


```suggestion
  src = fetchFromGitHub {
    owner = "swesterfeld";
    repo = "gst123";
    rev = version;
    hash = "";
  };
```

(update the hash.)

File: pkgs/applications/audio/gst123/default.nix [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  nativeBuildInputs = [ pkg-config wrapGAppsHook ];
```
Comment (AndersonTorres) [inline pkgs/applications/audio/gst123/default.nix:10]:
```suggestion
  nativeBuildInputs = [
    pkg-config
    wrapGAppsHook
  ];
```

File: pkgs/applications/audio/gst123/default.nix [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  buildInputs = [
```
Comment (AndersonTorres) [inline pkgs/applications/audio/gst123/default.nix:11]:
```suggestion

  buildInputs = [
```

File: pkgs/applications/audio/gst123/default.nix [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    ncurses
+  ] ++ (with gst_all_1; [
+    gstreamer
+    gst-plugins-base
+    gst-plugins-good
+    gst-plugins-bad
+    gst-plugins-ugly
+    gst-libav
+  ]);
+  meta = with lib; {
```
Comment (AndersonTorres) [inline pkgs/applications/audio/gst123/default.nix:22]:
```suggestion
  ]);

  meta = with lib; {
```

File: pkgs/applications/audio/gst123/default.nix [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    description = "GStreamer based command line media player";
+    homepage = "https://space.twc.de/~stefan/gst123.php";
+    license = licenses.lgpl2Plus;
+    maintainers = with maintainers; [ swesterfeld ];
+  };
```
Comment (AndersonTorres) [inline pkgs/applications/audio/gst123/default.nix:27]:
```suggestion
    inherit (ncurses.meta) platforms;
  };
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
Welcome to Nixpkgs! I'll be your guide.

File: pkgs/applications/audio/gst123/default.nix (hunk 3 above)
Comment (swesterfeld) [reply pkgs/applications/audio/gst123/default.nix:17]: That would break the build because gst123-0.4.0 assumes that it can access .git if it was build from git checkout. Nevertheless I looked at the code that needs .git, and its really legacy code for ChangeLog (which I think is no longer needed since developers usually have history via git), so since I'm upstream developer I'll go with your suggestion and make a gst123-0.4.1 release that doesn't need .git when building from git.

File: maintainers/maintainer-list.nix (hunk 1 above)
Comment (swesterfeld) [reply maintainers/maintainer-list.nix:16269]: Ok, I made this a separate commit and merged all your suggestions into one commit and force pushed the result.

File: pkgs/applications/audio/gst123/default.nix (hunk 3 above)
Comment (AndersonTorres) [reply pkgs/applications/audio/gst123/default.nix]: Usually it is not a good idea to rely on `.git`, since it is not a really stable interface.
Also, for us on Nixpkgs, it introduces too much impurity.

There are some flags we pass for those rare cases when `.git` is really needed, tho.

**Comment by AndersonTorres** [comment]
Result of `nixpkgs-review pr 248105` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>gst123</li>
  </ul>
</details>

File: pkgs/applications/audio/gst123/default.nix [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+  enableParallelBuilding = true;
+
+  meta = with lib; {
+    description = "GStreamer based command line media player";
+    homepage = "https://space.twc.de/~stefan/gst123.php";
+    license = licenses.lgpl2Plus;
+    maintainers = with maintainers; [ swesterfeld ];
+    inherit (ncurses.meta) platforms;
+  };
```
Comment (AndersonTorres) [inline pkgs/applications/audio/gst123/default.nix:49]: I have noticed in our CI it does not build on Apple/Darwin.
https://github.com/NixOS/nixpkgs/pull/248105/checks?check_run_id=15843117120

So, let's set it as broken on Darwin:

```suggestion
    inherit (ncurses.meta) platforms;
    broken = stdenv.isDarwin;
  };
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
Besides, LGTM

File: pkgs/applications/audio/gst123/default.nix [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+, fetchFromGitHub
+, autoreconfHook
+, pkg-config
+, wrapGAppsHook
+, gst_all_1
+, gtk3
+, ncurses
+}:
+
+stdenv.mkDerivation rec {
```
Comment (drupol) [inline pkgs/applications/audio/gst123/default.nix:12]: I think we should use the `finalAttrs` pattern here.

The `finalAttrs` pattern will let you remove the `rec` keyword ([see its implementation in Nix](https://github.com/NixOS/nix/blob/2d1d81114d72ace89ce08cd3bc93f4eb27a2975d/src/libexpr/nixexpr.cc#L121)).

Why is this pattern preferred to `rec` ?

Let's take this simple code example:

```nix
mkDerivation rec {
   foo = 1;
   bar = foo + 1;
}
```

and then `.overrideAttrs(old: { foo = 2; })`, you'll get `{ foo = 2; bar = 2; }` while with `finalAttrs` pattern, it would work correctly because it's a real fixed point.

Let me share a couple of useful links regarding the `finalAttrs` pattern:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent example of implementation: https://github.com/NixOS/nixpkgs/compare/17f96f7b978e61576cfe16136eb418f74fab9952..9e6ea843e473d34d4f379b8b0d8ef0425a06defe

Feel free to reach out if you need some assistance.

**Review by drupol — CHANGES_REQUESTED** [review]
Hi !

Thanks for your first contribution!

I've left some comments, let me know if this is clear enough.

File: pkgs/applications/audio/gst123/default.nix (hunk 9 above)
Comment (AndersonTorres) [reply pkgs/applications/audio/gst123/default.nix:12]: I believe we should not overload newcomers with this `finalAttrs` pattern yet. We can just modernize the expression later.

Comment (drupol) [reply pkgs/applications/audio/gst123/default.nix:12]: I did that with "newcomers" plenty of times and everything went well... why it would be an issue now?

Comment (AndersonTorres) [reply pkgs/applications/audio/gst123/default.nix:12]: OK then. 
Now, instead of saying "use finalAttrs, it is nice and cute", show the places where the code should be modified.

Comment (drupol) [reply pkgs/applications/audio/gst123/default.nix:12]: I've spent my time writing the explanation why this pattern should be used, with links. I tried to be complete so the new contributor is aware of what to do. Please, give them a chance to read and understand the reasoning behind this and come back if necessary. IMHO, I don't think that spotting each line where it should be changed is useful.

Comment (swesterfeld) [reply pkgs/applications/audio/gst123/default.nix]: Well, I am not really the right person to give a useful opinion on whether using `rec` or `finalAttrs` is better for new packages, but I changed it now.

**Review by AndersonTorres — APPROVED** [review]

**Review by drupol — APPROVED** [review]

**Comment by drupol** [comment]
Thanks for implementing all the changes, everything is just perfect now.

This PR will be merged as soon as the CI is done.

**Comment by drupol** [comment]
Oops. Going to submit a PR to fix this. Nice catch!
